ttnn/cpp/ttnn/operations/data_movement/common/common.cpp

```diff
@@ -55,6 +55,84 @@ ttnn::Tensor pad_to_tile_vol(
+
+    auto tensor_memory_layout = ttnn::TensorMemoryLayout::BLOCK_SHARDED;
+    if (strategy == ShardStrategy::WIDTH) {
+        tensor_memory_layout = ttnn::TensorMemoryLayout::WIDTH_SHARDED;
+    } else if (strategy == ShardStrategy::HEIGHT) {
+        tensor_memory_layout = ttnn::TensorMemoryLayout::HEIGHT_SHARDED;
+    }
+
+    auto shard_orientation = orientation;
+    auto shard_grid = core_grid;
+
+    auto height = shape[-2];
+    auto width = shape[-1];
+    std::array<uint32_t, 2> shard_shape;
+
+    if (use_height_and_width_as_shard_shape) {
+        if (shard_orientation == tt::tt_metal::ShardOrientation::ROW_MAJOR) {
+            shard_shape = {height, width};
+        } else if (shard_orientation == tt::tt_metal::ShardOrientation::COL_MAJOR) {
+            shard_shape = {width, height};
+        } else {
+            TT_THROW("Invalid shard orientation");
+        }
+    } else {
+        uint32_t batch_size = 1;
+        for (int i = 0; i < rank - 2; i++) {
+            batch_size *= shape[i];
+        }
+
+        auto tensor_height = batch_size * height;
+        auto tensor_width = width;
+        auto total_num_cores = shard_grid.num_cores();
+        auto grid_size = shard_grid.bounding_box().grid_size();
+
+        if (tensor_memory_layout == ttnn::TensorMemoryLayout::BLOCK_SHARDED) {
+            TT_ASSERT(grid_size.y * grid_size.x == total_num_cores, "Invalid CoreRangeSet for block sharding strategy");
+
+            if (shard_orientation == tt::tt_metal::ShardOrientation::ROW_MAJOR) {
+                auto tensor_height_padded =
+                    is_tile_layout ? tt::round_up(tensor_height, grid_size.y * 32) : tensor_height;
+                shard_shape = {tt::div_up(tensor_height_padded, grid_size.y), tt::div_up(tensor_width, grid_size.x)};
+            } else if (shard_orientation == tt::tt_metal::ShardOrientation::COL_MAJOR) {
+                auto tensor_height_padded =
+                    is_tile_layout ? tt::round_up(tensor_height, grid_size.x * 32) : tensor_height;
+                shard_shape = {tt::div_up(tensor_height_padded, grid_size.x), tt::div_up(tensor_width, grid_size.y)};
+            } else {
+                TT_THROW("Invalid shard orientation");
+            }
+        } else if (tensor_memory_layout == ttnn::TensorMemoryLayout::HEIGHT_SHARDED) {
+            auto tensor_height_padded = is_tile_layout ? tt::round_up(tensor_height, total_num_cores) : tensor_height;
+            shard_shape = {tt::div_up(tensor_height_padded, total_num_cores), tensor_width};
+        } else if (tensor_memory_layout == ttnn::TensorMemoryLayout::WIDTH_SHARDED) {
+            shard_shape = {tensor_height, tt::div_up(tensor_width, total_num_cores)};
+        } else {
+            TT_THROW("Invalid sharding scheme");
+        }
+    }
+
+    if (is_tile_layout && shard_shape[0] % 32 != 0 && shard_shape[1] % 32 != 0) {
+        TT_THROW("For sharding tiled tensors, the shard shape must fit neatly into tiles.");
```

Comment:
Can this exception be actually thrown? Doesn't this function pick the shard shape to be properly aligned already?

Reply:
also, reminder for error messages
https://github.com/tenstorrent/tt-metal/blob/main/contributing/ErrorMessageBestPractices.md